### PR TITLE
rewrite diagm with Pair's

### PR DESCRIPTION
**Pull request by fredrikekre**
ref JuliaLang/LinearAlgebra.jl#457, JuliaLang/julia#23757

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
`value => index` seems like the wrong ordering for this to me...

**Comment by fredrikekre** [comment]
Yes, I agree. Will update #23757 and then fix up this.

File: base/linalg/dense.jl
Comment (fredrikekre) [inline base/linalg/dense.jl]: Should we have this as `+=` to be consistent with how the `spdiagm` works?

File: NEWS.md
Comment (Sacha0) [inline NEWS.md]: Perhaps "pairs" -> "`Pair`s"?

File: base/linalg/dense.jl
Comment (Sacha0) [inline base/linalg/dense.jl]: The note re. `Diagonal` might be worth retaining (beyond merely the reference)?

File: base/linalg/dense.jl
Comment (Sacha0) [inline base/linalg/dense.jl]: Why `+=` rather than `=`?

File: base/linalg/generic.jl
Comment (Sacha0) [inline base/linalg/generic.jl]: `rank(diagm(0 => [1, 0, 2]))`?

File: test/linalg/bidiag.jl
Comment (Sacha0) [inline test/linalg/bidiag.jl]: These changes are a lovely illustration of how this API change improves things :).

**Review by Sacha0 — APPROVED** [review]
lgtm modulo the minor comments! Thanks @fredrikekre! :)

File: base/linalg/dense.jl
Comment (fredrikekre) [reply base/linalg/dense.jl]: To be consistent with `spdiagm`, since `combine` in `sparse` defaults to `+`. See https://github.com/JuliaLang/julia/pull/24047#discussion_r144005196 . Do you have another opinion about this?

Comment (Sacha0) [reply base/linalg/dense.jl]: Ah, do IIUC that its purpose is e.g. `[sp]diagm(0 => [1, 2, 3], 0 => [4, 5, 6])` yielding `[sp]diagm(0 => [5, 7, 9])`? If so, cheers, and thanks! :)

Comment (fredrikekre) [reply base/linalg/dense.jl]: Yes, I think it is the most natural thing to do.

**Comment by Sacha0** [comment]
Thanks @fredrikekre! :)

**Comment by andyferris** [comment]
I was wondering if `diagm(vector)` really should be deprecated. Intuitively I would guess that this function was primarily designed to work with... the diagonal. Typing `diagm(0 => vector)` seems slightly unnecessary in this basic use case. (My understanding was some of this work happened with the thought of deprecating `diagm` entirely, #23320, but if we have it then it may as well be easy to use).

**Comment by StefanKarpinski** [comment]
Agree: passing a single vector for the main diagonal seems perfectly legit to me, we should keep it.

**Pull request merged**
